integrations/access/servicenow/client.go

```diff
@@ -0,0 +1,312 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicenow
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/defaults"
+	"github.com/go-resty/resty/v2"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/integrations/lib"
+)
+
+const (
+	// DateTimeFormat is the time format used by servicenow
+	DateTimeFormat = "2006-01-02 15:04:05"
+)
+
+var (
+	incidentBodyTemplate = template.Must(template.New("incident body").Parse(
+		`{{.User}} requested permissions for roles {{range $index, $element := .Roles}}{{if $index}}, {{end}}{{ . }}{{end}} on Teleport at {{.Created.Format .TimeFormat}}.
+{{if .RequestReason}}Reason: {{.RequestReason}}{{end}}
+{{if .RequestLink}}To approve or deny the request, proceed to {{.RequestLink}}{{end}}
+`,
+	))
+	reviewNoteTemplate = template.Must(template.New("review note").Parse(
+		`{{.Author}} reviewed the request at {{.Created.Format .TimeFormat}}.
+Resolution: {{.ProposedState}}.
+{{if .Reason}}Reason: {{.Reason}}.{{end}}`,
+	))
+	resolutionNoteTemplate = template.Must(template.New("resolution note").Parse(
+		`Access request has been {{.Resolution}}
+{{if .ResolveReason}}Reason: {{.ResolveReason}}{{end}}`,
+	))
+)
+
+// Client is a wrapper around resty.Client.
+type Client struct {
+	ClientConfig
+
+	client *resty.Client
+}
+
+// ClientConfig is the config for the servicenow client.
+type ClientConfig struct {
+	// APIEndpoint is the endpoint for the Servicenow API
+	// api url of the form  https://instance.service-now.com/ with optional trailing '/'
+	APIEndpoint string
+
+	// WebProxyURL is the Teleport address used when building the bodies of the incidents
+	// allowing links to the access requests to be built
+	WebProxyURL *url.URL
+	// ClusterName is the name of the Teleport cluster
+	ClusterName string
+
+	// Username is the username used by the client for basic auth.
+	Username string
+	// APIToken is the token used for basic auth.
+	APIToken string
+}
+
+// NewClient creates a new Servicenow client for managing incidents.
+func NewClient(conf ClientConfig) (*Client, error) {
+	if err := conf.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	client := resty.NewWithClient(defaults.Config().HTTPClient)
+	client.SetBaseURL(conf.APIEndpoint).
+		SetHeader("Content-Type", "application/json").
+		SetHeader("Accept", "application/json").
+		SetBasicAuth(conf.Username, conf.APIToken)
+	return &Client{
+		client:       client,
+		ClientConfig: conf,
+	}, nil
+}
+
+func (conf ClientConfig) checkAndSetDefaults() error {
+	if conf.APIEndpoint == "" {
+		return trace.BadParameter("missing required field: APIEndpoint")
+	}
+	return nil
+}
```

Comment:
Are the remaining fields optional?

Reply:
For now yes. Once the oauth stuff is added in the next PR I'll add in checks that that is set.

Reply:
Username and Password shouldn't be optional I think